### DISCUSSION TITLE
Voice deserialization issue

### DISCUSSION THREAD

**Discussion by ChebanovDD** [RageAgainstThePixel/com.rest.elevenlabs#95] [Q&A] [answered]
# Bug Report

## Overview
I'm getting a `Newtonsoft.Json.JsonSerializationException: A member with the name 'name' already exists on 'ElevenLabs.Voices.Voice'. Use the JsonPropertyAttribute to specify another name.` exception when I try to get all voices:

```csharp
private ElevenLabsClient _client = new ElevenLabsClient(ApiKey);

public async Task<IEnumerable<Voice>> GetAllVoicesAsync(CancellationToken cancellationToken)
{
    return await _client.VoicesEndpoint.GetAllVoicesAsync(cancellationToken);
}
```

I can't reproduce the error when I create a new project, it only occurs in an existing project. I understand why it occurs, but I don't understand why it doesn't occur in a new project (maybe there is a difference in project configurations).

This is happening because the `JsonProperty` attribute is causing a conflict. Specifically, the JSON serializer is finding two properties with the same name, `name`, in the class `ElevenLabs.Voices.Voice`:

```csharp
[SerializeField]
private string name;

[Preserve]
[JsonProperty("name")]
public string Name
{
    get => name;
    private set => name = value;
}
```

The `Name` property is marked with `[JsonProperty("name")]`, which instructs Newtonsoft.Json to map the `Name` property to the `name` field in the JSON. However, the serializer also sees the private `name` field, which may cause it to believe there are two `name` fields in the class.

## Solution

**Rename the Private Fields** in the `ElevenLabs.Voices.Voice` and `ElevenLabs.Voices.VoiceSettings` classes, to avoid naming conflicts.

`ElevenLabs.Voices.Voice`:

```csharp
[SerializeField]
private string _name;

[Preserve]
[JsonProperty("name")]
public string Name
{
    get => _name;
    private set => _name = value;
}
```

`ElevenLabs.Voices.VoiceSettings`:

```csharp
[Range(0f, 1f)]
[SerializeField]
private float _stability = 0.75f;

[Preserve]
[JsonProperty("stability", DefaultValueHandling = DefaultValueHandling.Include)]
public float Stability
{
    get => _stability;
    set => _stability = value;
}

...

[Range(0f, 1f)]
[SerializeField]
private float _style = 0.45f;

[Preserve]
[JsonProperty("style", DefaultValueHandling = DefaultValueHandling.Include)]
public float Style
{
    get => _style;
    set => _style = value;
}
```

## Additional context
Using the `[JsonIgnore]` attribute doesn't works because the `[JsonIgnore]` only tells the JSON serializer to ignore the field when serializing or deserializing. However, the conflict occurs at the metadata level (how the serializer interprets the class structure), not during actual serialization. So, even if the private field is ignored during serialization, the serializer still sees two members with the same name `name`, leading to the conflict.

**Comment by StephenHodgson** [comment, marked as answer]
This shouldn't typically be a problem and since this package is well established, I will most likely not add the superfluous `_` prefix.
Even if I did rename the `FormerlySerializedAs("name")` attribute would still likely cause serialization errors similar to what you're seeing.

I have seen issues like this before in other project that somehow manage to silently rewrite the serialization options, 
do you have a custom serializer or something that changes the default Json serializer in your project?

One that comes to mind is https://github.com/applejag/Newtonsoft.Json-for-Unity.Converters

You'll need to disable the option to automatically serialize/deserialize private fields, or fix the project in question to properly use the NewtonSoft json instead of this package.

I dislike this package bc it rewrites the serialization options specifically set for the package in the Client.

**Reply by StephenHodgson** [reply]
See https://github.com/RageAgainstThePixel/com.openai.unity/discussions/171 since it is a similar discussion.

![image](https://github.com/user-attachments/assets/1cfdef1c-4c48-485c-9605-f107a9434de1)

**Comment by ChebanovDD** [comment]
Disabling `Use Unity Contract Resolver` solved the problem for me.

![image](https://github.com/user-attachments/assets/1a8603ea-3b66-406f-bd0f-c7de3c85e632)
